File: 581. Shortest Unsorted Continuous Subarray.cpp
int findUnsortedSubarray(vector<int>& nums) {
	int n = nums.size();
	if (n == 0 || n == 1) return 0;

	int f = 0;
	int e = 0;
	int min = INT_MAX;
	int max = INT_MIN;
	int temp = 0;
	// find the maximum not following ascending order
	for (int i = 1; i < n; ++i)
	{
		if (nums[i]<max)
		{
			e = i;
			continue;
		}
		temp = nums[i-1];
		if (nums[i] < temp&&max<temp)
		{
			max = temp;
			e = i;
		}
	}
	// find the minimum not following ascending order
	for (int i = n - 2; i >= 0; --i)
	{
		if (nums[i] > min)
		{
			f = i;
		}
		temp = nums[i+1];
		if (nums[i] > temp&& min > temp)
		{
			min = temp;
			f = i;
		}
	}
	// substract of the maximum and minimum is the answer
	return f==e ? 0 :e - f+1;
}

File: 581. Shortest Unsorted Continuous Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        q = []
        l = len(nums)
        # front
        for i in range(len(nums)):
            while len(q) > 0 and nums[q[-1]] > nums[i]:
                l = min(l, q.pop(-1))
            q.append(i)
        
        q = []
        r = 0
        for i in range(len(nums)-1, -1, -1):
            while len(q) > 0 and nums[q[-1]] < nums[i]:
                r = max(r, q.pop(-1))
            q.append(i)
        
        if l > r:
            return 0
        return r - l + 1

class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        xo = [i for i in nums]
        xo.sort()
        left = 0
        for i in range(len(nums)):
            if xo[i] != nums[i]:
                left = i
                break
        right = 0
        for i in range(len(nums)-1, -1, -1):
            if xo[i] != nums[i]:
                right = i
                break
        
        if right==0 and left == 0:
            return 0
        else:
            return right -left+1
